File: node/src/protocol.ts
/**
 * Protocol layer for forpc
 *
 * The wire format is a single protobuf-encoded `Packet` message. There is
 * no hand-written framing on top: encode/decode round-trip exclusively
 * through `Packet.encode` / `Packet.decode` from the generated protobuf
 * code, matching the Rust and Go implementations.
 *
 * Re-exports protobuf-generated `Call`, `Status`, `Packet` and `FrameKind`
 * symbols and provides small ergonomic helpers used by `peer.ts` /
 * `server.ts`.
 */

import proto from './generated/forpc.js'

const CallMessage = proto.forpc.Call
const StatusMessage = proto.forpc.Status
const PacketMessage = proto.forpc.Packet

// Frame kind constants, sourced from the generated protobuf enum.
export const FrameKind = proto.forpc.FrameKind

// gRPC-compatible status codes from protobuf definition
export const StatusCode = proto.forpc.StatusCode

export interface Call {
  method: string
  metadata: Record<string, string>
}

export interface Status {
  code: number
  message: string
}

export interface Packet {
  streamId: number
  kind: number
  payload: Buffer
  errorCode: number
}

/**
 * Encode a Call message to protobuf bytes
 */
export function encodeCall(call: Call): Buffer {
  const msg = CallMessage.create({
    method: call.method,
    metadata: call.metadata || {},
  })
  return Buffer.from(CallMessage.encode(msg).finish())
}

/**
 * Decode protobuf bytes to a Call message
 */
export function decodeCall(buf: Buffer): Call {
  const msg = CallMessage.decode(buf) as unknown as { method: string; metadata: Record<string, string> }
  return {
    method: msg.method || '',
    metadata: msg.metadata || {},
  }
}

/**
 * Encode a Status message to protobuf bytes
 */
export function encodeStatus(status: Status): Buffer {
  const msg = StatusMessage.create({
    code: status.code,
    message: status.message,
  })
  return Buffer.from(StatusMessage.encode(msg).finish())
}

/**
 * Decode protobuf bytes to a Status message
 */
export function decodeStatus(buf: Buffer): Status {
  const msg = StatusMessage.decode(buf) as unknown as { code: number; message: string }
  return {
    code: msg.code || 0,
    message: msg.message || '',
  }
}

/**
 * Encode a Packet to its protobuf wire bytes.
 */
export function encodePacket(packet: Packet): Buffer {
  const msg = PacketMessage.create({
    streamId: packet.streamId,
    kind: packet.kind,
    payload: packet.payload,
    errorCode: packet.errorCode,
  })
  return Buffer.from(PacketMessage.encode(msg).finish())
}

/**
 * Decode protobuf wire bytes into a Packet.
 */
export function decodePacket(data: Buffer): Packet {
  const msg = PacketMessage.decode(data) as unknown as {
    streamId: number
    kind: number
    payload: Uint8Array
    errorCode: number
  }
  return {
    streamId: msg.streamId || 0,
    kind: msg.kind || 0,
    payload: Buffer.from(msg.payload || []),
    errorCode: msg.errorCode || 0,
  }
}

/**
 * Create a HEADERS packet with an encoded Call
 */
export function headersPacket(streamId: number, call: Call): Packet {
  return {
    streamId,
    kind: FrameKind.HEADERS,
    payload: encodeCall(call),
    errorCode: 0,
  }
}

/**
 * Create a DATA packet
 */
export function dataPacket(streamId: number, payload: Buffer): Packet {
  return {
    streamId,
    kind: FrameKind.DATA,
    payload,
    errorCode: 0,
  }
}

/**
 * Create a TRAILERS packet with an encoded Status
 */
export function trailersPacket(streamId: number, status: Status): Packet {
  return {
    streamId,
    kind: FrameKind.TRAILERS,
    payload: encodeStatus(status),
    errorCode: 0,
  }
}

/**
 * Create a RST_STREAM packet with an error code carried in the dedicated
 * `errorCode` field of the protobuf message.
 */
export function rstStreamPacket(streamId: number, errorCode: number): Packet {
  return {
    streamId,
    kind: FrameKind.RST_STREAM,
    payload: Buffer.alloc(0),
    errorCode,
  }
}

/**
 * Create an OK status
 */
export function statusOk(): Status {
  return { code: StatusCode.OK, message: 'OK' }
}

File: go/forpc/protocol.go
package forpc

import (
	"google.golang.org/protobuf/proto"

	"github.com/bytemain/forpc/go/forpc/pb"
)

// Packet is the protobuf-defined wire packet (alias of pb.Packet) used by the
// peer to multiplex streams over a single transport. The wire format is
// produced and consumed entirely via protobuf encoding; there is no
// hand-written framing on top.
type Packet = pb.Packet

// FrameKind constants mirror the pb.FrameKind enum and are exposed under
// shorter names for convenience at call sites.
const (
	FrameHeaders   = pb.FrameKind_HEADERS
	FrameData      = pb.FrameKind_DATA
	FrameTrailers  = pb.FrameKind_TRAILERS
	FrameRstStream = pb.FrameKind_RST_STREAM
)

// EncodePacket marshals a Packet to its protobuf wire bytes.
func EncodePacket(p *Packet) ([]byte, error) {
	return proto.Marshal(p)
}

// DecodePacket parses a Packet from its protobuf wire bytes.
func DecodePacket(b []byte) (*Packet, error) {
	p := &Packet{}
	if err := proto.Unmarshal(b, p); err != nil {
		return nil, err
	}
	return p, nil
}

// HeadersPacket builds a HEADERS frame carrying an encoded Call.
func HeadersPacket(streamID uint32, call *pb.Call) (*Packet, error) {
	payload, err := proto.Marshal(call)
	if err != nil {
		return nil, err
	}
	return &Packet{StreamId: streamID, Kind: FrameHeaders, Payload: payload}, nil
}

// DataPacket builds a DATA frame carrying user payload bytes.
func DataPacket(streamID uint32, payload []byte) *Packet {
	return &Packet{StreamId: streamID, Kind: FrameData, Payload: payload}
}

// TrailersPacket builds a TRAILERS frame carrying an encoded Status.
func TrailersPacket(streamID uint32, status *pb.Status) (*Packet, error) {
	payload, err := proto.Marshal(status)
	if err != nil {
		return nil, err
	}
	return &Packet{StreamId: streamID, Kind: FrameTrailers, Payload: payload}, nil
}

// RstStreamPacket builds a RST_STREAM frame with the given error code.
func RstStreamPacket(streamID uint32, errorCode uint32) *Packet {
	return &Packet{StreamId: streamID, Kind: FrameRstStream, ErrorCode: errorCode}
}

// Call is a peer-side view of a pb.Call decoded from a HEADERS frame.
type Call struct {
	Method   string
	Metadata map[string]string
}

// Status is a peer-side view of a pb.Status decoded from a TRAILERS frame.
type Status struct {
	Code    pb.StatusCode
	Message string
}

func StatusOKValue() Status {
	return Status{Code: pb.StatusCode_OK, Message: "OK"}
}

func (s Status) IsOK() bool {
	return s.Code == pb.StatusCode_OK
}
